### PR TITLE
CB-20071: Use the Rolling OS API in case of Rolling Upgrade

### PR DESCRIPTION
**Pull request by tiborpopovics**
There is a new API where you have to tell the order of the instances to be replaced with a new OS. In case of Rolling Upgrade

### CODE REVIEW

**Comment by keyki** [comment]
@sodre90 @lnardai please review - this is for MD DL Rolling OS upgrade

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    private static final String CORE_GROUP = "core";
+
+    private static final String AUXILIARY_GROUP = "auxiliary";
+
+    private static final String IDBROKER_GROUP = "idbroker";
+
+    private static final String GATEWAY_GROUP = "gateway";
+
+    public OrderedOSUpgradeSetRequest createOrderedOSUpgradeSetRequest(StackV4Response stackV4Response) {
```
Comment (sodre90) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java:35]: One minor thing: maybe we can rename it to something like createMediumDutyOrderedOSUpgradeSetRequest?

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        LOGGER.debug("Creating OrderedOSUpgradeSetRequest for rolling OS upgrade");
+        Map<String, List<InstanceMetaDataV4Response>> instanceMetaDataByInstanceGroup = getInstanceMetaDataByInstanceGroup(stackV4Response);
+        Map<String, List<String>> instanceIdsByInstanceGroup = getInstanceIdsByInstanceGroup(instanceMetaDataByInstanceGroup);
+        LOGGER.debug("Instance ids by instance group: {}", instanceIdsByInstanceGroup);
+
```
Comment (sodre90) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java:40]: I think we should validate node counts in groups

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        List<OrderedOSUpgradeSet> osUpgradeByUpgradeSets = new ArrayList<>();
+        osUpgradeByUpgradeSets.add(new OrderedOSUpgradeSet(0, Set.of(
```
Comment (sodre90) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java:42]: Should we deal with AZs?

**Review by sodre90 — CHANGES_REQUESTED** [review]

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java (hunk 3 above)
Comment (tiborpopovics) [reply datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java:42]: We discussed with @akanto and @doktoric  that this configuration is fine for multi AZ as well.

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.StackV4Response;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.response.instancegroup.instancemetadata.InstanceMetaDataV4Response;
+import com.sequenceiq.cloudbreak.common.exception.CloudbreakServiceException;
+
+@Component
+public class OrderedOSUpgradeRequestProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(OrderedOSUpgradeRequestProvider.class);
+
+    private static final String MASTER_GROUP = "master";
```
Comment (pdarvasi) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java:25]: how about using an enum for this?

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                        entry -> entry.getValue().stream()
+                                .map(InstanceMetaDataV4Response::getInstanceId)
+                                .collect(Collectors.toList())));
+    }
+
+    private String getInstanceId(Map<String, List<String>> instanceIdsByInstanceGroup, String instanceGroup) {
+        try {
+            return instanceIdsByInstanceGroup.get(instanceGroup).remove(0);
+        } catch (Exception e) {
+            String message = String.format("There are no instances left in the %s group. Remaining instances: %s", instanceGroup, instanceIdsByInstanceGroup);
```
Comment (pdarvasi) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java:84]: Could you pls include a hint for the error cause here, too?

**Review by pdarvasi — CHANGES_REQUESTED** [review]

**Pull request merged**